### PR TITLE
Reflect Validation Script

### PR DESCRIPTION
**Pull request by brianjohnson5972**
**Change Description**

Added a script to validate that FC_REFLECT and FC_REFLECT_ENUM macro uses are consistent with their class definitions.

**Consensus Changes**

N/A

**API Changes**

N/A

**Documentation Additions**

N/A (This is a script for internal development and not for general use, so the documentation in the file is sufficient)

### CODE REVIEW

File: tools/validate_reflection.py [hunk 1]
```diff
@@ -0,0 +1,713 @@
...
+    all=match.group(0)
+    debug("singleline found=%s" % (all))
+    if EmptyScope.single_comment_ignore_swap_pattern.match(all):
+        return all
+    else:
+        return "\n"
+
+def validate_file(file):
+    f = open(file, "r")
+    contents = f.read()
```
Comment (jgiszczak) [inline tools/validate_reflection.py]: `contents = '\n' + f.read() # lazy fix for complex regex`

Strings being immutable, Python probably copies the entire file contents twice as it was.

File: tools/validate_reflection.py [hunk 2]
```diff
@@ -0,0 +1,713 @@
...
+    found = re.search(fc_reflect_str, contents)
+    if found is None:
+        return
+    print("validate %s" % (file))
+    debug("validate %s" % (file))
+    global_namespace=Namespace("", None, 0, contents, None)
+    global_namespace.read()
+    if args.debug:
+        _, filename = os.path.split(file)
+        f = open(temp_dir + "/" + filename + ".struct", "w")
```
Comment (jgiszczak) [inline tools/validate_reflection.py]: Replacing this and next two lines:
```python
with open(os.path.join(temp_dir, filename + '.struct'), 'w') as f:
    f.write("global_namespace=%s" % (global_namespace))
```
`close()` becomes implicit and will be executed even if there's an exception during `write()`.

File: tools/validate_reflection.py [hunk 3]
```diff
@@ -0,0 +1,713 @@
...
+        fwd_swapped = []
+        ignored = []
+        f_index = 0
+        f_len = len(class_struct.fields)
+        while f_index < f_len:
+            field = class_struct.fields[f_index]
+            reflect_field = reflection.fields[rf_index] if rf_index < rf_len else None
+            processed.append(field)
+            debug("\nfield=%s reflect_field=%s" % (field, reflect_field))
+            #debug("start rf_index=%s, rf_len=%s, f_index=%s, f_len=%s" % (rf_index, rf_len, f_index, f_len))
```
Comment (jgiszczak) [inline tools/validate_reflection.py]: Commented code.

File: tools/validate_reflection.py [hunk 4]
```diff
@@ -0,0 +1,713 @@
...
+        unused_reflect_fields = []
+        while rf_index < rf_len:
+            debug("rf_index=%s, rf_len=%s fields=%s" % (rf_index, rf_len, ",".join(reflection.fields)))
+            reflect_field = reflection.fields[rf_index]
+            if reflect_field in back_swapped:
+                back_swapped.remove(reflect_field)
+            else:
+                unused_reflect_fields.append(reflect_field)
+            rf_index += 1
+        assert len(unused_reflect_fields) == 0, "Reflection for %s has fields not in defintion for class/struct - \"%s\"" % (reflection_name, ",".join(unused_reflect_fields))
```
Comment (jgiszczak) [inline tools/validate_reflection.py]: definition

File: tools/validate_reflection.py [hunk 5]
```diff
@@ -0,0 +1,713 @@
...
+def walk(current_dir):
+    result = True
+    print("Searching for files: %s" % (current_dir))
+    for root, dirs, filenames in os.walk(current_dir):
+        for filename in filenames:
+            _, extension = os.path.splitext(filename)
+            if extension not in extensions:
+                continue
+            try:
+                validate_file(root + "/" + filename)
```
Comment (jgiszczak) [inline tools/validate_reflection.py]: `validate_file(os.path.join(root, filename))`

File: tools/validate_reflection.py [hunk 6]
```diff
@@ -0,0 +1,713 @@
...
+                f_index += 1
+                continue
+            if reflect_field in reflection.swapped:
+                debug("field \"%s\" swapped (fwd)" % (field))
+                reflection.swapped.remove(reflect_field)
+                fwd_swapped.append(reflect_field)
+                assert reflect_field in reflection.fields, "Reflection for %s indicates swapping field %s but it doesn't exist in that class/struct so it should be removed" % (reflection_name, reflect_field)
+                rf_index += 1
+                continue
+            assert reflect_field not in ignored, "Reflection for %s should not indicate %s for %s, it should indicate %s - %s" % (reflection_name, ignore_str, reflect_field, swap_str, ",".join(ignored))
```
Comment (jgiszczak) [inline tools/validate_reflection.py]: Semicolon or period in the message, not comma.

File: tools/validate_reflection.py [hunk 7]
```diff
@@ -0,0 +1,713 @@
...
+parser.add_argument('-x', '--extension', type=str, help="extensions array to allow for directory and recursive search.  Defaults to \".hpp\" and \".cpp\".", action='append')
+parser.add_argument('-e', '--exit-on-error', help="Exit immediately when a validation error is discovered.  Default is to run validation on all files and directories provided.", action='store_true')
+parser.add_argument('files', metavar='file', nargs='+', type=str, help="File containing nodes info in JSON format.")
+args = parser.parse_args()
+
+recurse = args.recurse
+if args.debug:
+    if not os.path.exists(temp_dir):
+        os.mkdir(temp_dir)
+    debug_file = open(temp_dir + "/validate_reflect.debug", "w")
```
Comment (jgiszczak) [inline tools/validate_reflection.py]: `    debug_file = open(os.path.join(temp_dir, 'validate_reflect.debug', 'w')`

File: tools/validate_reflection.py [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+#    Looks for files with FC_REFLECT macros.  Requires fields to match class definition (provided in same file),
+#  unless the FC_REFLECT is proceeded by "// @ignore <field1>, <field2>, ..., <fieldN>" to indicate that field1,
+#  field2, ... fieldN are not reflected and/or "// @swap <field1>, <field2>, ..., <fieldN>" to indicate that
+#  field1, field2, ... fieldN are not in the same order as the class definition.
+#
+#  NOTE:  If swapping fields the script expects you to only indicate fields that are not in the expected order,
+#  so once it runs into the swapped field, it will remove that field from the order and expect the remaining in
+#  that order, so if the class has field1, field2, field3, and field4, and the reflect macro has the order
+#  field1, field3, field2, then field4, it should indicate swapping field2.  This will remove field2 from the
+#  expected order and the rest will now match.  Or else it should indicated swapping field3, since the remaining
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py [hunk 9]
```diff
@@ -0,0 +1,713 @@
...
+#  fields will also match the order.  But both field2 and field3 should not be indicated.
+#
+#
+# 
+###############################################################
+
+
+temp_dir = "temp_dir"
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py [hunk 10]
```diff
@@ -0,0 +1,713 @@
...
+parser.add_argument('-d', '--debug', help="generate debug output into \"./%s\"" % (temp_dir), action='store_true')
+parser.add_argument('-r', '--recurse', help="recurse through an entire directory (if directory provided for \"file\"", action='store_true')
+parser.add_argument('-x', '--extension', type=str, help="extensions array to allow for directory and recursive search.  Defaults to \".hpp\" and \".cpp\".", action='append')
+parser.add_argument('-e', '--exit-on-error', help="Exit immediately when a validation error is discovered.  Default is to run validation on all files and directories provided.", action='store_true')
+parser.add_argument('files', metavar='file', nargs='+', type=str, help="File containing nodes info in JSON format.")
+args = parser.parse_args()
+
+recurse = args.recurse
+if args.debug:
+    if not os.path.exists(temp_dir):
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py [hunk 11]
```diff
@@ -0,0 +1,713 @@
...
+        os.mkdir(temp_dir)
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py [hunk 12]
```diff
@@ -0,0 +1,713 @@
...
+    debug_file = open(temp_dir + "/validate_reflect.debug", "w")
+else:
+    debug_file = None
+extensions = []
+if args.extension is None or len(args.extension) == 0:
+    extensions = [".hpp",".cpp"]
+else:
+    for extension in args.extension:
+        assert len(extension) > 0, "empty --extension passed in"
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Args.extension is an appended string, so it is a list of string, so I'm verifying that no empty strings are passed, but glad to remove that check if you don't think it is useful.

File: tools/validate_reflection.py [hunk 13]
```diff
@@ -0,0 +1,713 @@
...
+    def __init__(self, name, start, content, parent_scope):
+        pname = parent_scope.name if parent_scope is not None else ""
+        debug("EmptyScope.__init__ %s %d - Parent %s" % (name, start, pname))
+        self.name = name
+        self.content = content
+        self.start = start
+        self.current = start + 1
+        self.parent_scope = parent_scope
+        self.end = len(content) - 1 if start == 0 else None
+        self.start_char = "{"
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py [hunk 14]
```diff
@@ -0,0 +1,713 @@
...
+        self.end_char = "}"
+        self.children = {}
+        self.children_ordered = []
+        self.fields = []
+        self.usings = {}
+        self.type = "EmptyScope"
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Type is set in each constructor, and used for debugging, so it has to be on each object.

File: tools/validate_reflection.py [hunk 15]
```diff
@@ -0,0 +1,713 @@
...
+        loc = content.find(find_str, start, end)
+        if loc == -1:
+            return loc
+        else:
+            return loc + len(find_str) - len(self.start_char)
+
+    def find_possible_end(self):
+        possible = self.content.find(self.end_char, self.current)
+        if possible == -1:
+            possible = len(self.content) - 1
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py [hunk 16]
```diff
@@ -0,0 +1,713 @@
...
+        return Namespace(name, inherit, start, content, parent_scope)
+    elif type == EmptyScope.class_str or type == EmptyScope.struct_str:
+        return ClassStruct(name, inherit, start, content, parent_scope, is_enum = False)
+    elif type == EmptyScope.enum_str:
+        return ClassStruct(name, inherit, start, content, parent_scope, is_enum = True)
+    else:
+        assert False, "Script does not account for type = \"%s\" found in \"%s\"" % (type, content[start:])
+
+class ClassStruct(EmptyScope):
+   #field_pattern = re.compile(r'\n\s*?(?:mutable\s+)?(\w[\w:\d<>]+)\s*(\w+)\s*(?:=\s[^;]+)?;', re.MULTILINE | re.DOTALL)
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py [hunk 17]
```diff
@@ -0,0 +1,713 @@
...
+    cb_obj_pattern = re.compile(r'chainbase::object$')
+    obj_pattern = re.compile(r'^object$')
+    using_pattern = re.compile(r'\n\s*?using\s+(\w+)\s*=\s*([\w:]+)(?:<.*>)?;')
+
+    def __init__(self, name, inherit, start, content, parent_scope, is_enum):
+        debug("ClassStruct.__init__ %s %d" % (name, start))
+        EmptyScope.__init__(self, name, start, content, parent_scope)
+        self.classes = {}
+        self.pattern = ClassStruct.class_pattern
+        self.type = "ClassStruct"
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Same as above

File: tools/validate_reflection.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jgiszczak) [reply tools/validate_reflection.py]: Instances of an object can access their class constants, and all instances of `class EmptyScope` have `self.type = "EmptyScope"`.  But the whole thing can be dispensed with.  Eliminate `self.type` and use `self.__class__.__name__`, which is the same thing but created by the interpreter so it's always correct even if the class gets renamed.

File: tools/validate_reflection.py [hunk 18]
```diff
@@ -0,0 +1,713 @@
...
+        self.ignored = []
+        self.swapped = []
+        self.absent = []
+
+class Reflections:
+    def __init__(self, content):
+        self.content = content
+        self.current = 0
+        self.end = len(content)
+        self.classes = {}
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py [hunk 19]
```diff
@@ -0,0 +1,713 @@
...
+        if invalid_chars:
+            for ic in invalid_chars.groups():
+                debug("invalid_char=%s" % (ic))
+            debug("WARNING: looks like \"%s\" is intending to %s, but there are invalid characters - \"%s\"" % (all, ignore_or_swap, ",".join(invalid_chars.groups())))
+            return ""
+        groups=re.findall(EmptyScope.multi_line_comment_ignore_swap_pattern, all)
+        if groups is None:
+            return ""
+        rtn_str="// %s " % (ignore_or_swap)
+        add=False
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py [hunk 20]
```diff
@@ -0,0 +1,713 @@
...
+        for group in groups:
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py (hunk 1 above)
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py (hunk 2 above)
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py (hunk 3 above)
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

File: tools/validate_reflection.py (hunk 6 above)
Comment (brianjohnson5972) [reply tools/validate_reflection.py]: Ok

**Pull request merged**